# What Prices the Car - Analysing Prices of Cars in Ebay

We'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. 
The dataset was originally scraped and uploaded to Kaggle. You can get the data here https://www.kaggle.com/orgesleka/used-cars-database/data. 

Objective is to clean the data and study the included used car listings to analyze what are determinants of the pricing of the car. The data columns are as the following:-

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

As there are a number of columns, we can explore dropping some of the columns which are redundant.

In [1]:
import pandas as pd
import numpy as np

#read in the data
autos = pd.read_csv('autos.csv',encoding  = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Let's explore the values within the columns here. 

In [5]:
name = autos["name"].unique()
name

array(['Peugeot_807_160_NAVTECH_ON_BOARD',
       'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik',
       'Volkswagen_Golf_1.6_United', ...,
       'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon',
       'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!',
       'Fiat_500_C_1.2_Dualogic_Lounge'], dtype=object)

In [6]:
seller = autos["seller"].unique()
seller

array(['privat', 'gewerblich'], dtype=object)

In [7]:
offertype = autos["offerType"].unique()
offertype

array(['Angebot', 'Gesuch'], dtype=object)

In [8]:
price = autos["price"].unique()
price

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [9]:
price.max()

'$999,999'

In [10]:
price.min()

'$0'

In [11]:
abtest = autos["abtest"].unique()
abtest

array(['control', 'test'], dtype=object)

In [12]:
vehicletype = autos["vehicleType"].unique()
vehicletype

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [13]:
model = autos["model"].unique()
model

array(['andere', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'fiesta',
       'wrangler', 'clubman', 'a6', 'transporter', 'astra', 'v40',
       'ibiza', 'micra', '1er', 'yaris', 'colt', '6_reihe', '5_reihe',
       'corolla', 'ka', 'tigra', 'punto', 'vito', 'cordoba', 'galaxy',
       '100', 'sharan', 'octavia', 'm_klasse', 'lupo', 'superb', 'meriva',
       'c_max', 'laguna', 'touran', '1_reihe', 'm_reihe', 'touareg',
       'seicento', 'avensis', 'vivaro', 'x_reihe', 'ducato', 'carnival',
       'boxster', 'signum', 'zafira', 'rav', 'a5', 'beetle', 'c_reihe',
       'phaeton', 'i_reihe', 'sl', 'insignia', 'up', 'civic', '80',
       'mx_reihe', 'omega', 'sorento', 'z_reihe', 'berlingo', 'clk',
       '

In [14]:
fueltype = autos["fuelType"].unique()
fueltype

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [15]:
nrofpictures = autos["nrOfPictures"].unique()
nrofpictures

array([0])

We can make the following observation of the data
- columns with missing infos : when there is complete data it is 5000 records. The following columns are missing info - vehicleType, gearbox, model, fueltype, notrepaireddamage
- column headers : it is a mix of lower and upper case. We will need to tidy this up
- price : it is not in float / integer. We will need to convert it for calculation
- nrofpictures : number of pictures is only zero. However we won't be working with images in this project
- odometer : it includes the string 'km'. We will need to tidy up this data by removing the trailing string 'km'
- columns with almost identical data : seller, offertype

# Cleaning Column Names

We'll start by converting the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive. The use of snakecase is prescribed in Python's Style Guide https://www.python.org/dev/peps/pep-0008/. Another reason is to make it more readable. 

In [16]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [17]:
def clean_col(col):
    col = col.replace("yearOfRegistration","registration_year")
    col = col.replace("monthOfRegistration","registration_month")
    col = col.replace("notRepairedDamage","unrepaired_damage")
    col = col.replace("dateCreated","ad_created")
    col = col.replace("dateCrawled","date_crawled")
    col = col.replace("offerType","offer_type")
    col = col.replace("vehicleType","vehicle_type")
    col = col.replace("powerPS","power_ps")
    col = col.replace("fuelType","fuel_type")
    col = col.replace("nrOfPictures","nr_of_pictures")
    col = col.replace("postalCode","postal_code")
    col = col.replace("lastSeen","last_seen")
    return col

In [18]:
new_columns = []

for row in autos.columns:
    clean_c = clean_col(row)
    new_columns.append(clean_c)
new_columns

['date_crawled',
 'name',
 'seller',
 'offer_type',
 'price',
 'abtest',
 'vehicle_type',
 'registration_year',
 'gearbox',
 'power_ps',
 'model',
 'odometer',
 'registration_month',
 'fuel_type',
 'brand',
 'unrepaired_damage',
 'ad_created',
 'nr_of_pictures',
 'postal_code',
 'last_seen']

In [19]:
autos.columns = new_columns

In [20]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Now the column headers are cleaned, changed the case from camel case to snake case as following Python convention in PEP8.

# Initial Exploration and Cleaning

We will look for text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. We will also look for numeric data stored as text which can be cleaned and converted.

In [21]:
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

There are some observations to be made here. 
1. registration_year min is 1000 - this is unlikely as car is invented in the 19th century. 
2. power_ps max is 17700 much higher than 75th percentile - this indicates there might be outliers skewing the data distribution
3. nr_of_pictures only have 1 value - this is a candidate to drop the column.

In [22]:
#Changing Price to Numeric type. We'll start by removing the dollar and comma symbol
autos["price"] = autos["price"].str.replace('$','')
autos["price"] = autos["price"].str.replace(',','')

In [23]:
print(autos["price"].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: object


In [24]:
autos["price"] = autos["price"].astype(int)
print(autos["price"].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


In [25]:
#Changing Price to Numeric type. We'll start by removing km and comma symbol
autos["odometer"] = autos["odometer"].str.replace('km','')
autos["odometer"] = autos["odometer"].str.replace(',','')

print(autos["odometer"].head())

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: object


In [26]:
autos["odometer"] = autos["odometer"].astype(int)
print(autos["odometer"].head())

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64


In [27]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

In [28]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

# Exploring the Odometer (Mileage) and Price Columns

##### Explore Odometer Columns

In [29]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

From here, can be observed that a number of the cars has high mileage. 

In [30]:
print(autos["price"].unique().shape)
print("\n")
print(autos["price"].describe())
print("\n")
print(autos["price"].value_counts().head(20))

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64


In [31]:
autos["price"].value_counts().tail(20)

910       1
238       1
2671      1
69900     1
151990    1
2479      1
4510      1
86500     1
47499     1
16998     1
27299     1
41850     1
4780      1
686       1
6495      1
20790     1
8970      1
846       1
2895      1
33980     1
Name: price, dtype: int64

In [32]:
autos["price"].value_counts().sort_index(ascending=True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

In [33]:
autos["price"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

There are cars with the price 0 which are about 1421. On the other extreme, there are cars priced at more than 1million, which looks quite high. From a quick search on the Internet, prices of cars in Germany is around $36K.  

Given that this is an auction site, the prices may climb, so let's cap it at $500,000. 

In [34]:
autos = autos[autos["price"].between(1,500000)]

In [35]:
print(autos["price"].unique().shape)
print("\n")
print(autos["price"].describe())
print("\n")
print(autos["price"].value_counts().head(20))

(2346,)


count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


500     781
1500    734
2500    643
1200    639
1000    639
600     531
3500    498
800     498
2000    460
999     434
750     433
900     420
650     419
850     410
700     395
4500    394
300     384
2200    382
950     379
1100    376
Name: price, dtype: int64


In [36]:
print(autos["odometer_km"].unique().shape )
print("\n")
print(autos["odometer_km"].describe())
print("\n")
print(autos["odometer_km"].value_counts().head(3))

(13,)


count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    31414
125000     5057
100000     2115
Name: odometer_km, dtype: int64


In [37]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

# Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself.
- `date_crawled`
- `last_seen`
- `ad_created`
- `registration_month`
- `registration_year`

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

##### Exploring date_crawled column

In [38]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [39]:
autos["date_crawled"].value_counts(normalize=True, dropna=False)

2016-03-19 17:36:18    0.000062
2016-03-09 11:54:38    0.000062
2016-03-11 22:38:16    0.000062
2016-03-30 19:48:02    0.000062
2016-04-02 11:37:04    0.000062
2016-03-25 19:57:10    0.000062
2016-03-12 16:06:22    0.000062
2016-03-21 20:37:19    0.000062
2016-04-04 16:40:33    0.000062
2016-04-02 15:49:30    0.000062
2016-03-14 20:50:02    0.000062
2016-03-23 19:38:20    0.000062
2016-03-08 10:40:35    0.000062
2016-03-29 23:42:13    0.000062
2016-03-22 09:51:06    0.000062
2016-03-23 18:39:34    0.000062
2016-03-16 21:50:53    0.000062
2016-03-05 16:57:05    0.000062
2016-03-28 18:36:36    0.000041
2016-03-21 15:57:37    0.000041
2016-03-17 18:47:24    0.000041
2016-04-04 17:06:17    0.000041
2016-03-15 16:53:56    0.000041
2016-03-12 15:44:10    0.000041
2016-03-06 14:47:11    0.000041
2016-03-20 12:37:40    0.000041
2016-03-12 19:43:05    0.000041
2016-03-23 18:57:02    0.000041
2016-03-08 11:55:06    0.000041
2016-03-20 13:58:29    0.000041
                         ...   
2016-03-

In [40]:
autos["date_crawled"].sort_index()

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
5        2016-03-21 13:47:45
6        2016-03-20 17:55:21
7        2016-03-16 18:55:19
8        2016-03-22 16:51:34
9        2016-03-16 13:47:02
10       2016-03-15 01:41:36
11       2016-03-16 18:45:34
12       2016-03-31 19:48:22
13       2016-03-23 10:48:32
14       2016-03-23 11:50:46
15       2016-04-01 12:06:20
16       2016-03-16 14:59:02
17       2016-03-29 11:46:22
18       2016-03-26 19:57:44
19       2016-03-17 13:36:21
20       2016-03-05 19:57:31
21       2016-03-06 19:07:10
22       2016-03-28 20:50:54
23       2016-03-10 19:55:34
24       2016-04-03 11:57:02
25       2016-03-21 21:56:18
26       2016-04-03 22:46:28
28       2016-03-19 21:56:19
29       2016-04-02 12:45:44
30       2016-03-14 11:47:31
                ...         
49968    2016-04-01 17:49:15
49969    2016-03-17 18:49:02
49970    2016-03-21 22:47:37
49971    2016-

Most of the date crawled is on Sun, Mon, Sat in that order. 

##### Exploring ad created

In [41]:
autos["ad_created"].describe()

count                   48565
unique                     76
top       2016-04-03 00:00:00
freq                     1887
Name: ad_created, dtype: object

In [42]:
autos["ad_created"].value_counts(normalize=True, dropna=False)

2016-04-03 00:00:00    0.038855
2016-03-20 00:00:00    0.037949
2016-03-21 00:00:00    0.037579
2016-04-04 00:00:00    0.036858
2016-03-12 00:00:00    0.036755
2016-03-14 00:00:00    0.035190
2016-04-02 00:00:00    0.035149
2016-03-28 00:00:00    0.034984
2016-03-07 00:00:00    0.034737
2016-03-29 00:00:00    0.034037
2016-03-15 00:00:00    0.034016
2016-03-19 00:00:00    0.033687
2016-04-01 00:00:00    0.033687
2016-03-30 00:00:00    0.033501
2016-03-08 00:00:00    0.033316
2016-03-09 00:00:00    0.033151
2016-03-11 00:00:00    0.032904
2016-03-22 00:00:00    0.032801
2016-03-26 00:00:00    0.032266
2016-03-23 00:00:00    0.032060
2016-03-10 00:00:00    0.031895
2016-03-31 00:00:00    0.031875
2016-03-25 00:00:00    0.031751
2016-03-17 00:00:00    0.031278
2016-03-27 00:00:00    0.030989
2016-03-16 00:00:00    0.030125
2016-03-24 00:00:00    0.029280
2016-03-05 00:00:00    0.022897
2016-03-13 00:00:00    0.017008
2016-03-06 00:00:00    0.015320
                         ...   
2016-02-

Most of the ads-created are on Sunday, Monday, Saturday in that order.

##### Exploring Last Seen column

In [43]:
autos["last_seen"].describe()

count                   48565
unique                  38474
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object

In [44]:
autos["last_seen"].value_counts(normalize=True, dropna=False)

2016-04-07 06:17:27    0.000165
2016-04-07 03:16:17    0.000144
2016-04-06 21:17:51    0.000144
2016-04-07 06:46:12    0.000124
2016-04-06 14:17:04    0.000124
2016-04-07 04:46:51    0.000124
2016-04-06 15:45:50    0.000124
2016-04-06 15:16:45    0.000124
2016-04-06 02:17:26    0.000124
2016-04-07 05:16:17    0.000124
2016-04-07 03:45:23    0.000124
2016-04-06 02:16:12    0.000124
2016-04-06 10:17:12    0.000124
2016-04-06 20:48:27    0.000124
2016-04-06 13:17:03    0.000124
2016-04-06 08:44:19    0.000124
2016-04-06 07:46:11    0.000124
2016-04-06 22:17:26    0.000124
2016-04-05 16:44:47    0.000124
2016-04-06 15:17:56    0.000124
2016-04-06 05:16:14    0.000124
2016-04-06 06:17:24    0.000124
2016-04-06 14:44:55    0.000124
2016-04-06 19:16:38    0.000124
2016-04-06 22:45:50    0.000103
2016-04-05 15:17:49    0.000103
2016-04-07 04:15:59    0.000103
2016-04-05 20:46:01    0.000103
2016-04-05 18:44:55    0.000103
2016-04-05 23:46:00    0.000103
                         ...   
2016-03-

For Last Seen, values of the highest frequency are on days Wed, Thu, Tue

##### Exploring Registration Year

In [45]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [46]:
autos["registration_year"].head()

0    2004
1    1997
2    2009
3    2007
4    2003
Name: registration_year, dtype: int64

In [47]:
autos["registration_year"].value_counts(normalize=True, dropna=False)

2000    0.064985
2005    0.060455
1999    0.059652
2004    0.055657
2003    0.055575
2006    0.054978
2001    0.054278
2002    0.051189
1998    0.048656
2007    0.046886
2008    0.045609
2009    0.042932
1997    0.040173
2011    0.033419
2010    0.032719
2017    0.028663
1996    0.028271
2012    0.026974
1995    0.025265
2016    0.025121
2013    0.016535
2014    0.013652
1994    0.012952
2018    0.009678
1993    0.008751
2015    0.008072
1992    0.007619
1990    0.007145
1991    0.006980
1989    0.003583
          ...   
1950    0.000062
9999    0.000062
1800    0.000041
1951    0.000041
1941    0.000041
1934    0.000041
1954    0.000041
1957    0.000041
1955    0.000041
2019    0.000041
1001    0.000021
1939    0.000021
4500    0.000021
1953    0.000021
1111    0.000021
4800    0.000021
5911    0.000021
1943    0.000021
1938    0.000021
1929    0.000021
2800    0.000021
6200    0.000021
4100    0.000021
8888    0.000021
1927    0.000021
9000    0.000021
1948    0.000021
1000    0.0000

In [48]:
autos["registration_year"].sort_index()

0        2004
1        1997
2        2009
3        2007
4        2003
5        2006
6        1995
7        1998
8        2000
9        1997
10       2017
11       2000
12       2010
13       1999
14       2007
15       1982
16       1999
17       1990
18       1995
19       2004
20       2003
21       2004
22       2015
23       2010
24       2014
25       1996
26       1992
28       2007
29       2004
30       2002
         ... 
49968    1986
49969    2005
49970    2010
49971    2001
49972    2004
49973    2004
49975    2012
49976    1992
49977    2003
49978    1996
49979    2011
49980    1995
49981    1998
49982    2004
49983    1999
49985    1995
49986    2010
49987    2013
49988    2001
49989    1997
49990    2012
49991    2016
49992    2009
49993    1997
49994    2001
49995    2011
49996    1996
49997    2014
49998    2013
49999    1996
Name: registration_year, Length: 48565, dtype: int64

Majority of the cars are registered in 1999 - 2005. There are a few cars registered in 1000 and 9000!!!

# Dealing with Incorrect Registration Year Data

For registration_year, the acceptable value would be between 1900 - 2016. As the scraping happened in 2016, any registration above 2016 would be inaccurate. 

In [49]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [50]:
autos["registration_year"].value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002035
          ...   
1966    0.000471
1976    0.000450
1969    0.000407
1975    0.000386
1965    0.000364
1964    0.000257
1963    0.000171
1959    0.000129
1961    0.000129
1910    0.000107
1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1951    0.000043
1934    0.000043
1957    0.000043
1955    0.000043
1953    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.000021
1948    0.000021
1927    0.0000

In [51]:
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

# Exploring Price by Brand

In [52]:
brand_unique = autos["brand"].unique()
brand_unique

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [53]:
brand_describe = autos["brand"].describe()

In [54]:
brand_describe

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

In [55]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

The most popular brands are Volkswagen, BMW, Opel. As there are a lot of brands that don't have significant percentge of listings, we will limit our analysis to brands representing more than 5% of total listings. 

In [56]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [57]:
# Create an empty dictionary to store the results
common_avg_price_by_brand = {}

# Create an array of unique brands
brand_unique = autos["brand"].unique()

# Use a for loop to iterate over the brands
for b in common_brands:
    # Use boolean comparison to select only rows that
    # correspond to a specific brand
    selected_rows = autos[autos["brand"] == b]
    # Calculate the mean average price for just those rows
    mean = selected_rows["price"].mean()
    # Assign the mean value to the dictionary, using the
    # brand name as the key
    common_avg_price_by_brand[b] = mean

In [58]:
common_avg_price_by_brand

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287}

From the above, there is a distinct price gap between the brands. Audi, Mercedes_Benz, BMW is much more expensive than Ford, Opel. 

# Storing Aggregate Data in a DataFrame

From the above, let us calculate the mean mileage for each of the top brands to see if there is any correlation between mileage and pricing. To understand if there is linkage between these 2 dimensions, we need to combine the data into a single dataframe (with a shared index) and display the dataframe directly. We will use pandas series constructor, and pandas dataframe constructor to achieve this goal. 

In [59]:
# Create an empty dictionary to store the results
common_avg_mileage_by_brand = {}

# Use a for loop to iterate over the brands
for b in common_avg_price_by_brand:
    # Use boolean comparison to select only rows that
    # correspond to a specific brand
    selected_rows = autos[autos["brand"] == b]
    # Calculate the mean average mileage for just those rows
    mean = selected_rows["odometer_km"].mean()
    # Assign the mean value to the dictionary, using the
    # brand name as the key
    common_avg_mileage_by_brand[b] = mean

In [60]:
common_avg_mileage_by_brand

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056}

In [61]:
#Convert avg_price_by_brand dictionaries to series objects, using the series constructor
bmp_series = pd.Series(common_avg_price_by_brand)

In [62]:
bmp_series

volkswagen       5402.410262
bmw              8332.820518
opel             2975.241935
mercedes_benz    8628.450366
audi             9336.687454
ford             3749.469507
dtype: float64

In [63]:
#Convert avg_mileage_by_brand dictionaries to series objects, using the series constructor
bmm_series = pd.Series(common_avg_mileage_by_brand)

In [64]:
bmm_series

volkswagen       128707.158791
bmw              132572.513140
opel             129310.035842
mercedes_benz    130788.363313
audi             129157.386785
ford             124266.012872
dtype: float64

In [65]:
#Create a dataframe from bmp series object using the dataframe constructor
autos_agg_brand = pd.DataFrame(bmp_series, columns=['mean_price'])

In [66]:
autos_agg_brand

mean_price
volkswagen     5402.410262
bmw            8332.820518
opel           2975.241935
mercedes_benz  8628.450366
audi           9336.687454
ford           3749.469507

In [67]:
autos_agg_brand['mean_mileage'] = bmm_series

In [68]:
autos_agg_brand

mean_price   mean_mileage
volkswagen     5402.410262  128707.158791
bmw            8332.820518  132572.513140
opel           2975.241935  129310.035842
mercedes_benz  8628.450366  130788.363313
audi           9336.687454  129157.386785
ford           3749.469507  124266.012872

In [69]:
autos_agg_brand['price_over_mileage'] = autos_agg_brand['mean_price'] / autos_agg_brand['mean_mileage']

In [70]:
autos_agg_brand.sort_values(by=['mean_price'])

mean_price   mean_mileage  price_over_mileage
opel           2975.241935  129310.035842            0.023009
ford           3749.469507  124266.012872            0.030173
volkswagen     5402.410262  128707.158791            0.041974
bmw            8332.820518  132572.513140            0.062855
mercedes_benz  8628.450366  130788.363313            0.065973
audi           9336.687454  129157.386785            0.072289

In [71]:
autos_agg_brand.sort_values(by=['mean_mileage'])

mean_price   mean_mileage  price_over_mileage
ford           3749.469507  124266.012872            0.030173
volkswagen     5402.410262  128707.158791            0.041974
audi           9336.687454  129157.386785            0.072289
opel           2975.241935  129310.035842            0.023009
mercedes_benz  8628.450366  130788.363313            0.065973
bmw            8332.820518  132572.513140            0.062855

From here, the range of car mileage does not vary much as the prices by brands. Cars which are more expensive has a bit of higher mileage, whilst the less expensive cars have lower mileage. 

# Find the most common brand/model combinations

In [72]:
autos['brand_model'] = autos['brand'] + ['_'] + autos['model']

In [73]:
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps  ... odometer_km  registration_month  fuel_type  \
0    manuell       158  ...      150000                   3        lpg   
1  automatik       286  ...      150000                   6     benzin   
2    manuell       102  ...       70000                   7     benzin   
3  automatik        71  ...       70000                   6     benzin   
4    manuell         0  ...      150000                   7     benzin   

        brand unrepaired_damage           ad_created nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00              0   
1         bmw              nein  2016-04-04 00:00:00              0   
2  volkswagen              nein  2016-03-26 00:00:00              0   
3       smart              nein  2016-03-12 00:00:00              0   
4        ford              nein  2016-04-01 00:00:00              0   

   postal_code            last_seen      brand_model  
0        79588  2016-04-06 06:45:54   peugeot_andere  
1        71034  2016-04-06 14:45:08          bmw_7er  
2        35394  2016-04-06 20:15:37  volkswagen_golf  
3        33729  2016-03-15 03:16:28     smart_fortwo  
4        39218  2016-04-01 14:38:50       ford_focus  

[5 rows x 21 columns]

In [74]:
#instead of frequency table, can just use a value count for dataframe
autos.brand_model.value_counts()

volkswagen_golf                  3707
bmw_3er                          2615
volkswagen_polo                  1609
opel_corsa                       1592
volkswagen_passat                1349
opel_astra                       1348
audi_a4                          1231
mercedes_benz_c_klasse           1136
bmw_5er                          1132
mercedes_benz_e_klasse            958
audi_a3                           825
audi_a6                           797
ford_focus                        762
ford_fiesta                       722
volkswagen_transporter            674
renault_twingo                    615
peugeot_2_reihe                   600
smart_fortwo                      550
opel_vectra                       544
mercedes_benz_a_klasse            539
bmw_1er                           521
ford_mondeo                       479
renault_clio                      473
mercedes_benz_andere              439
volkswagen_touran                 433
fiat_punto                        415
opel_zafira 

The highest 3 cars brands with models are Volkswagen Golf, BMW 3er, Volkswagen Polo. 

# Split the odometer_km into groups

We want to see if prices are affected by the mileage of cars listed

In [75]:
autos.groupby(pd.cut(autos["odometer_km"], np.arange(50000, 110000, 10000))).mean()

price  registration_year    power_ps  odometer_km  \
odometer_km                                                                 
(50000, 60000]   12385.004433        2006.881206  125.865248      60000.0   
(60000, 70000]   10927.182814        2006.218197  132.184499      70000.0   
(70000, 80000]    9721.947636        2005.927273  130.242182      80000.0   
(80000, 90000]    8465.025105        2004.801554  119.216975      90000.0   
(90000, 100000]   8132.697279        2003.868805  117.889699     100000.0   

                 registration_month  nr_of_pictures   postal_code  
odometer_km                                                        
(50000, 60000]             6.171986             0.0  51489.656028  
(60000, 70000]             6.155855             0.0  53670.871946  
(70000, 80000]             6.197818             0.0  51630.339636  
(80000, 90000]             5.943216             0.0  52163.072325  
(90000, 100000]            5.933431             0.0  52429.017007

From here, it is clear that price drops steadily as mileage increases. only the price column is applicable. 

# How much cheaper are cars with damage than their non-damaged counterparts

In [76]:
non_repaired = autos.loc[(autos["unrepaired_damage"] == "ja"), ["brand", "model", "price", "unrepaired_damage"]]
print("non_repaired cars price distribution \n \n", non_repaired["price"].describe(),"\n \n")

non_repaired cars price distribution 
 
 count     4540.000000
mean      2241.146035
std       3563.276478
min          1.000000
25%        500.000000
50%       1000.000000
75%       2500.000000
max      44200.000000
Name: price, dtype: float64 
 



In [77]:
repaired = autos.loc[(autos["unrepaired_damage"] == "nein"), ["brand", "model", "price", "unrepaired_damage"]]
print("repaired cars price distribution \n \n", repaired["price"].describe(), "\n \n")

repaired cars price distribution 
 
 count     33834.000000
mean       7164.033103
std       10078.475478
min           1.000000
25%        1800.000000
50%        4150.000000
75%        9000.000000
max      350000.000000
Name: price, dtype: float64 
 



If we look at the above, the mean of the cars that are not repaired are more than 50% lower than a repaired car damage. 

# Summary

In summary, what we can conclude is that the most popular cars here are German made - Volkswagen, BMW, Opel. Secondly, we can also analyse and derive from here that cars prices drops as mileage increase. Another observation is that damaged cars that are not repaired are more than 50% cheaper than damaged cars that are repaired. 